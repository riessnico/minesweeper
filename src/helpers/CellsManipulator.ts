import { Coords, Field } from "./Field";

export const incrementNeighbours = (coords: Coords, field: Field) => {
	return field
}