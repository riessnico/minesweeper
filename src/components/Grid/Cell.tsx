import styled from '@emotion/styled';
import React from 'react';
import { Cell as CellType, CellState, Coords } from '@/helpers/Field';
import { useMouseDown } from '@/hooks/useMouseDown';

export interface CellProps {
  /**
   * Cell status based on cellType
   */
  children: CellType;
  /**
   * Cell coordinates
   */
  coords: Coords;
  /**
   * onClick by cell handler
   */
  onClick: (coords: Coords) => void;
  /**
   * onContextMeny by cell handler
   */
  onContextMenu: (coords: Coords) => void;
}

export const checkCellIsActive = (cell: CellType) =>
  [CellState.hidden, CellState.flag, CellState.weakFlag].includes(cell);

export const CellComponent: React.FC<CellProps> = ({
  children,
  coords,
  ...rest
}) => {
  const [mouseDown, onMouseDown, onMouseUp] = useMouseDown();

  const onClick = () => rest.onClick(coords);

  const onContextMenu = (elem: React.MouseEvent<HTMLElement>) => {
    /**
     * Prevent context menu by default
     */
    elem.preventDefault();

    if (checkCellIsActive(children)) {
      rest.onContextMenu(coords);
    }
  };

  const props = {
    onClick,
    onContextMenu,
    'data-testid': `${children}_${coords}`,
    onMouseDown,
    onMouseUp,
    onMouseLeave: onMouseUp,
    mouseDown,
  };

  return <ComponentsMap {...props}>{children}</ComponentsMap>;
};

interface ComponentsMapProps {
  children: CellType;
  onClick: (elem: React.MouseEvent<HTMLDivElement>) => void;
  onContextMenu: (elem: React.MouseEvent<HTMLDivElement>) => void;
  onMouseDown: () => void;
  onMouseUp: () => void;
  onMouseLeave: () => void;
  mouseDown: boolean;
  'data-testid'?: string;
}

const ComponentsMap: React.FC<ComponentsMapProps> = ({ children, ...rest }) => {
  const nonActiveCellProps = {
    onContextMenu: rest.onContextMenu,
    'data-testid': rest['data-testid'],
  };

  switch (children) {
    case CellState.empty:
      return <RevealedFrame {...nonActiveCellProps} />;
    case CellState.bomb:
      return (
        <BombFrame {...nonActiveCellProps}>
          <Bomb />
        </BombFrame>
      );
    case CellState.flag:
      return (
        <ClosedFrame {...rest}>
          <Flag />
        </ClosedFrame>
      );
    case CellState.weakFlag:
      return (
        <ClosedFrame {...rest}>
          <WeakFlag />
        </ClosedFrame>
      );
    case CellState.hidden:
      return <ClosedFrame {...rest} />;
    default:
      return <RevealedFrame {...nonActiveCellProps}>{children}</RevealedFrame>;
  }
};

interface ClosedFrameProps {
  mouseDown?: boolean;
}

export const ClosedFrame = styled.div<ClosedFrameProps>`
  display: flex;
  align-items: center;
  justify-content: center;
  user-select: none;
  cursor: pointer;
  width: 1.8vw;
  height: 1.8vw;
  background-color: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: ${({ mouseDown = false }) =>
    mouseDown ? 'transparent' : 'white #e9e9e9 #e9e9e9 white'};
  &:hover {
    filter: brightness(1.05);
  }
`;

const transparent = 'rgba(0,0,0,0)';
const colors: { [key in CellType]: string } = {
  0: '#000',
  1: '#2a48ec',
  2: '#2bb13d',
  3: '#ec6561',
  4: '#233db7',
  5: '#a6070f',
  6: '#e400af',
  7: '#906a02',
  8: '#fa0707',
  9: transparent,
  10: transparent,
  11: transparent,
  12: transparent,
};

const RevealedFrame = styled(ClosedFrame)`
  border-color: #dddddd;
  cursor: default;
  color: ${({ children }) => colors[children as CellType] ?? transparent};
  &:hover {
    filter: brightness(1);
  }
`;

const Bomb = styled.div`
  border-radius: 50%;
  width: 1vh;
  height: 1vh;
  background-color: #333;
`;

const BombFrame = styled(RevealedFrame)`
  background-color: #ec433c;
`;

const Flag = styled.div`
  width: 0px;
  height: 0px;
  border-top: 0.5vh solid transparent;
  border-bottom: 0.5vh solid transparent;
  border-left: 0.5vh solid #ec433c;
`;

const WeakFlag = styled(Flag)`
  border-left: 0.5vw solid #f19996;
`;
