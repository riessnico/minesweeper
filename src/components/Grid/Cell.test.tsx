import { CellState, Coords } from '@/helpers/Field';
import { createEvent, fireEvent, render, screen } from '@testing-library/react';
import React from 'react';
import { CellComponent as Cell, checkCellIsActive } from './Cell';

describe('Cell component check', () => {
  const coords: Coords = [1, 1];

  for (let cell = CellState.empty; cell <= CellState.weakFlag; cell++) {
    it('Check prevent default contextMenu for every type of cell', () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${cell}_${coords}`);
      console.log(cellComp);
      const contextMenuEvent = createEvent.contextMenu(cellComp);
      fireEvent(cellComp, contextMenuEvent);

      expect(contextMenuEvent.defaultPrevented).toBe(true);
    });

    it('onClick and onContextMenu handler should be called for active cells', () => {
      const props = {
        coords,
        onClick: jest.fn(),
        onContextMenu: jest.fn(),
      };

      render(<Cell {...props}>{cell}</Cell>);

      const cellComp = screen.getByTestId(`${cell}_${coords}`);
      fireEvent.click(cellComp);
      fireEvent.contextMenu(cellComp);
      if (checkCellIsActive(cell)) {
        expect(props.onClick).toBeCalled();
        expect(props.onContextMenu).toBeCalled();
      } else {
        expect(props.onClick).not.toBeCalled();
        expect(props.onContextMenu).not.toBeCalled();
      }
    });
  }
});
