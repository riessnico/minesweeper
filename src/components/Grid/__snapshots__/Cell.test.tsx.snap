// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell component check Cell renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-srqhig"
    data-testid="0_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-bffs1t"
    data-testid="1_1,1"
  >
    1
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-b3qosu"
    data-testid="2_1,1"
  >
    2
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-1bynhd7"
    data-testid="3_1,1"
  >
    3
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-1ox0h26"
    data-testid="4_1,1"
  >
    4
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-p6m5uo"
    data-testid="5_1,1"
  >
    5
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-1w514aa"
    data-testid="6_1,1"
  >
    6
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-sxfluf"
    data-testid="7_1,1"
  >
    7
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-v58ylb"
    data-testid="8_1,1"
  >
    8
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-1tk1181"
    data-testid="9_1,1"
  >
    <div
      class="css-81z0jo"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-1d1wied"
    data-testid="10_1,1"
  />
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-1d1wied"
    data-testid="11_1,1"
  >
    <div
      class="css-186ts50"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Cell renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-1d1wied"
    data-testid="12_1,1"
  >
    <div
      class="css-wx1uuw"
    />
  </div>
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 1`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 2`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 3`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 4`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 5`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 6`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 7`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 8`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 9`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 10`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 11`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 12`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;

exports[`Cell component check Closed Frame renders correctly 13`] = `
<DocumentFragment>
  <div
    class="css-bx680z"
  />
</DocumentFragment>
`;
